client2/thin.go

```diff
@@ -0,0 +1,308 @@
+
+	ID *[MessageIDLength]byte
+
+	// Payload is the decrypted payload plaintext.
+	Payload []byte
+}
+
+// ThinClient is the client that handles communication between the mixnet application
+// and the client daemon. It does not do any encryption or decryption or checking
+// of cryptographic signatures; those responsibilities are left to the client daemon
+// process.
+type ThinClient struct {
+	worker.Worker
+
+	log          *log.Logger
+	unixConn     *net.UnixConn
+	destUnixAddr *net.UnixAddr
+
+	pkidoc      *cpki.Document
+	pkidocMutex sync.RWMutex
+
+	isConnected bool
+
+	receivedCh chan ThinResponse
+}
+
+// NewThinClient creates a new thing client.
+func NewThinClient() *ThinClient {
+	return &ThinClient{
+		log: log.NewWithOptions(os.Stderr, log.Options{
+			Prefix: "thin_client",
+			Level:  log.DebugLevel,
+		}),
+		receivedCh: make(chan ThinResponse),
+	}
+}
+
+// Close halts the thin client worker thread and closes the socket
+// connection with the client daemon.
+func (t *ThinClient) Close() error {
+	err := t.unixConn.Close()
+	t.Halt()
+	return err
+}
+
+// Dial dials the client daemon via our agreed upon abstract unix domain socket.
+func (t *ThinClient) Dial() error {
+	t.log.Debug("Dial begin")
+	srcUnixAddr, err := net.ResolveUnixAddr("unixpacket", "@katzenpost_golang_thin_client")
+	if err != nil {
+		return err
+	}
+
+	t.destUnixAddr, err = net.ResolveUnixAddr("unixpacket", "@katzenpost")
+	if err != nil {
+		return err
+	}
+
+	t.log.Debugf("Dial unixpacket %s %s", srcUnixAddr, t.destUnixAddr)
+	t.unixConn, err = net.DialUnix("unixpacket", srcUnixAddr, t.destUnixAddr)
+	if err != nil {
+		return err
+	}
+
+	// WAIT UNTIL we have a Noise cryptographic connection with an edge node
+	t.log.Debugf("Waiting for a connection status message")
+	message1, err := t.readNextMessage()
+	if err != nil {
+		return err
+	}
+	if !message1.IsStatus {
+		panic("did not receive a connection status message")
+	}
+	if !message1.IsConnected {
+		return errors.New("not connected")
+	}
+
+	t.log.Debugf("Waiting for a PKI doc message")
+	message2, err := t.readNextMessage()
+	if err != nil {
+		return err
+	}
+	t.parsePKIDoc(message2.Payload)
+	t.Go(t.worker)
+	t.log.Debug("Dial end")
+	return nil
+}
+
+func (t *ThinClient) worker() {
+	for {
+		select {
+		case <-t.HaltCh():
+			return
+		default:
+		}
+
+		message, err := t.readNextMessage()
+		if err != nil {
+			t.log.Infof("thin client ReceiveMessage failed: %v", err)
+		}
+		if message == nil {
+			return
+		}
+
+		t.log.Debug("THIN CLIENT WORKER RECEIVED A MESSAGE")
+
+		switch {
+		case message.IsStatus == true:
+			t.isConnected = message.IsConnected
+		case message.IsPKIDoc == true:
+			t.parsePKIDoc(message.Payload)
+		default:
+			if message.Payload == nil {
+				t.log.Infof("message.Payload is nil")
+			}
+			response := ThinResponse{
+				SURBID:  message.SURBID,
+				ID:      message.ID,
+				Payload: message.Payload,
+			}
+			select {
+			case <-t.HaltCh():
```

Comment:
should  we print to debug log that we are shutting down the worker? (same in `select {` above)